torch/utils/_python_dispatch.py

```diff
@@ -151,7 +151,99 @@ def transform_subclass(t, callback):
+           if isinstance(ret, torch._subclasses.functional_tensor.FunctionalTensor):
+               # directly calling this overload, and passing ret.shape, because we **explicitly**
+               # don't want to reset the sizes on ret, if the storage implies a size change.
+               # Why?
+               # - FunctionalTensor has a CustomSizes, so that it can plumb all calls to metadata to the inner tensor.
+               # - TensorImpl.set_sizes_strides() therefore yells if you try to directly change sizes.
+               torch.ops.aten.set_.source_Storage_storage_offset(ret, arg.untyped_storage(), 0, ret.shape)
+           else:
+               ret.set_(arg.untyped_storage())
+
+    def is_match(arg, ret):
+        arg_aliases = set() if not arg.alias_info else arg.alias_info.before_set
+        out_aliases = set() if not ret.alias_info else ret.alias_info.before_set
+        return len(arg_aliases & out_aliases) > 0
+
+    num_args = len(func._schema.arguments)
+    num_returns = len(func._schema.returns)
+    for arg_idx in range(num_args):
+        for return_idx in range(num_returns):
+            if is_match(func._schema.arguments[arg_idx], func._schema.returns[return_idx]):
+                alias_storage(args[arg_idx], outs[return_idx])
+
+    # Sigh... the torchscript parser has a bug where alias annotations for Tensor[](a) don't show up properly
+    # See https://github.com/pytorch/pytorch/issues/106173
+    if func.overloadpacket in [
+        torch.ops.aten.chunk,
+        torch.ops.aten.tensor_split,
+        torch.ops.aten.split,
+        torch.ops.aten.split_with_sizes,
+        torch.ops.aten.hsplit,
+        torch.ops.aten.vsplit,
+        torch.ops.aten.dsplit,
+        torch.ops.aten.unbind,
+    ]:
+        assert isinstance(outs, list) and all(isinstance(x, torch.Tensor) for x in outs)
+        for o in outs:
+            # For lists of outputs, need to alias every individual tensor to the input
+            alias_storage(args[0], o)
+
+def return_and_correct_aliasing(func, args, out):
+    def get_write_alias(x):
+        if not x.alias_info or not x.alias_info.before_set:
+            return None
+        before_set = list(x.alias_info.before_set)
+        # torchscript allows for complicated alias sets, but our dispatcher ops only really involve simple aliasing
+        assert len(before_set) == 1
+        if '!' in list(before_set)[0]:
+            return before_set[0]
+        return None
+
+    # Fix up the storages of any outs so that they point to the same storage as the input,
+    # if func is a view op.
+    _correct_storage_aliasing(func, args, [out] if not isinstance(out, (list, tuple)) else out)
+
+    # Next: we need to make sure to return inputs directly, if the output is a mutable alias (e.g. add_()).
+
+    # simple case: none of our outputs have mutable aliases, so we can return the output as-is
+    if not any(get_write_alias(r) is not None for r in func._schema.returns):
+        return out
```

Comment:
Just to make sure I understand what's going on: Let's say we have `sin_(Tensor(a!) x) -> Tensor(a!)`. It looks like we:
1. set the storage on the output to be the same as the input
2. then attempt to return the input
Is that correct? (or does `_correct_storage_aliasing` do something to not set the storage?)

Reply:
Coming back to this late, sorry (now that proxy/fake portions of this PR stack is landed, I'm starting to rework the python functionalization PR's).

`_correct_storage_aliasing()` will be a no-op, unless `func` is a view op (its output is a non-mutable alias of an input).

So for an op like `sin_()`, only (2) will happen above. For an op like `aten.view()`, only (1) will happen above.

It's probably technically possible to have to worry about an opt that returns **both** an output that aliases an input, **and** an input itself, but that would probably not work in PT2 for other reasons (functionalization probably wouldn't be able to handle such an op by default)